GeneratorPgn.py

```diff
@@ -0,0 +1,273 @@
+        """
+        while True:
+            position = self.get_position()
+            self.replay_queue.put(position)
+
+
+class IteratorPgn(object):
+    def __init__(
+        self,
+        batch_size: int,
+        generators_paths,
+        replay_size=30000,
+        replay_initial=10000,
+        n_generators=16,
+    ):
+        """
+        :param batch_size batch_size: size of batch yielded from replay buffer
+        :param batch_size generators_paths: a list containing sets of paths for individual generators
+        :param batch_size replay_size: size of replay buffer
+        :param batch_size replay_initial: samples to initialize buffer before yielding
+        :param batch_size n_generators: how many generators should generate data
+        """
+        if type(generators_paths) == str:
+            n_generators = 1
+        elif len(generators_paths) < n_generators:
+            n_generators = len(generators_paths)
+        self.replay_queue = mp.Queue(maxsize=n_generators * 4)
+        self.generators = []
+        for paths in generators_paths:
+            self.generators.append(GeneratorPgn(self.replay_queue, paths))
+        self.replay_buffer = deque(maxlen=replay_size)
+        self.batch_size = batch_size
+        self.replay_size = replay_size
+        self.replay_initial = replay_initial
+        self.active_processes = []
+
+    def kill_processes(self):
+        """
+        Stops executing all processes triggered by this class.
+        :return: None
+        """
+        for process in self.active_processes:
+            process.terminate()
+            process.join()
+        self.active_processes.clear()
+
+    def initialize_processes(self, kill_old=True):
+        """
+        Starts executing processes that generate data, kills old processes if necessary.
+        :param kill_old: it `True` then kills old processes, default: True
+        :return: None
+        """
+        if kill_old:
+            self.kill_processes()
+        for generator in self.generators:
+            process = mp.Process(target=generator.play_func, args=())
+            process.start()
+            self.active_processes.append(process)
+
+    def __iter__(self):
+        """
+        Iterator function that samples batches from generated data and initializes processes.
+        :return: yields batches of samples
+        """
+        try:
+            self.initialize_processes()
+            while True:
+                while self.replay_queue.qsize():
+                    entry = self.replay_queue.get()
+                    self.replay_buffer.append(entry)
+                if len(self.replay_buffer) < self.replay_initial:
+                    continue
+                yield random.sample(self.replay_buffer, self.batch_size)
+        except Exception:
+            pass
+        finally:
+            self.kill_processes()
+            
+    first_cords = {'a': 0, 'b': 1, 'c': 2, 'd': 3, 'e': 4, 'f': 5, 'g': 6, 'h': 7}

```

Comment:
this indentation is suspicious, first_cords is now in the GeneratorPgn class - this should not work

Reply:
it's true i have already corrected it